### PR TITLE
fix(quic): don't report error for gracefully closed connections

### LINKED ISSUE #4588: quic: returning ListenerClosed on connection closed internally
## Current behaviour

In lighthouse while trying to shutdown the client, [SwarmEvent::Listenerclosed](https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/lighthouse_network/src/service/mod.rs#L1535-L1544) was returned and logged:
```
CRIT Listener closed reason: Err(Custom { kind: Other, error: Transport(Custom { kind: Other, error: Right(EndpointDriverCrashed) }) }), addresses: [\"/ip4/127.0.0.1/udp/9001/quic-v1\", \"/ip4/141.95.85.227/udp/9001/quic-v1\"], service: libp2p
```
In the `quic` transport, when polling the `Listener` :
https://github.com/libp2p/rust-libp2p/blob/399eaa39156f756956f38fab20083293da826fdd/transports/quic/src/transport.rs#L573-L595

as `quinn` (unlike tokio's `TcpStream` which cannot be closed) [`Accept`](https://docs.rs/quinn/0.10.2/quinn/struct.Accept.html) `Future` impl is `Option`, it yields `None` when the endpoint has been closed and `quic::Transport` returns `SwarmEvent::ListenerClosed`.

This is different from the other transports, namely `tcp` which only returns `SwarmEvent::ListenerClosed` when `Transport::remove_listener` is called.

## Possible Solution
Can we not send `SwarmEvent::ListenerClosed` when `Accept` returns `None` and just return `Poll::Ready(None)` to match other transports? @kpp what was the reason for considering `Accept` returning `None` a `EndpointDriverCrashed` ?

thanks
## Version

- libp2p version (version number, commit, or branch):

## Would you like to work on fixing this bug?

Yes

### PR DESCRIPTION
**Pull request by jxs**
## Description

Closes quinn's `Endpoint` with `Ok(())`  when `Accept` returns `None`. 

Resolves: #4588.
Related: https://github.com/quinn-rs/quinn/issues/1676.
## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/quic/CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,10 @@
 ## 0.9.3 - unreleased
 
+- Close quinn's `Endpoint` with `Ok(())` when `Accept` returns `None`. See [PR 4621].
```
Comment (mxinden) [inline transports/quic/CHANGELOG.md]: The changelog is written for users, not for maintainers. I think this is a good git commit message. I think for a changelog you could write something along the lines of:

> No longer report error when explicit closing of a QUIC endpoint succeeds.

**Review by mxinden — APPROVED** [review]
Thank you for following this all the way through. Much appreciated.

Small suggestions for the changelog entry. Otherwise this looks good to me.

File: transports/quic/src/lib.rs [hunk 2]
```diff
@@ -90,10 +90,6 @@ pub enum Error {
     #[error(transparent)]
     Io(#[from] std::io::Error),
 
-    /// The task to drive a quic endpoint has crashed.
-    #[error("Endpoint driver crashed")]
-    EndpointDriverCrashed,
```
Comment (thomaseizinger) [inline transports/quic/src/lib.rs]: Removing this is a breaking change. Can you just deprecate the variant to say that it is no longer emitted and we remove it in the next breaking release?

File: transports/quic/CHANGELOG.md (hunk 1 above)
Comment (thomaseizinger) [reply transports/quic/CHANGELOG.md]:
```suggestion
- No longer report error when explicit closing of a QUIC endpoint succeeds.
  See [PR 4621].
```

Comment (jxs) [reply transports/quic/CHANGELOG.md]: yeah makes sense, thanks guys!

File: transports/quic/src/lib.rs (hunk 2 above)
Comment (jxs) [reply transports/quic/src/lib.rs]: yup, did not remember that, thanks Thomas! Addressed

**Review by thomaseizinger — APPROVED** [review]
Thanks @jxs!

**Comment by jxs** [comment]
Btw, `kad` tests are flaky: https://github.com/libp2p/rust-libp2p/actions/runs/6489236865/job/17623145967?pr=4621#step:6:765

**Pull request merged**